### PR TITLE
Refactoring ExtensionCompilerPass

### PR DESCRIPTION
**Pull request by core23**
## Subject

Minor refactoring to prepare #6027.

I am targeting this branch, because this is BC.

### CODE REVIEW

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 1]
```diff
@@ -156,7 +136,22 @@ protected function getExtensionsForAdmin($id, Definition $admin, ContainerBuilde
...
+            throw new RuntimeException(sprintf('The admin "%s" does not have a valid manager', $admin->getClass()));
+        }
+
+        if (!\is_string($class)) {
+            throw new \InvalidArgumentException(sprintf(
+                'Invalid argument "%s" for admin class "%s',
+                $argument,
+                $admin->getClass()
+            ));
+        }
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:152]:
```suggestion
        if (!\is_string($class)) {
            throw new \TypeError(sprintf(
                'Argument "%s" for admin class "%s" must be of type string, %s given.',
                $argument,
                $admin->getClass(),
                \is_object($class) ? \get_class($class) : \gettype($class)
            ));
        }
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 2]
```diff
@@ -214,6 +209,32 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
...
+            }
+        } elseif ('extends' === $type) {
+            if ($classReflection->isSubclassOf($subject)) {
+                return true;
+            }
+        } elseif ('uses' === $type) {
+            if ($this->hasTrait($classReflection, $subject)) {
+                return true;
+            }
+        }
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]:
```suggestion
        if ('instanceof' === $type && ($classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName())) {
            return true;
        }
        if ('implements' === $type && $classReflection->implementsInterface($subject)) {
            return true;
        }
        if ('extends' === $type && $classReflection->isSubclassOf($subject)) {
            return true;
        }
        if ('uses' === $type && $this->hasTrait($classReflection, $subject)) {
            return true;
        }
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 3]
```diff
@@ -96,13 +97,14 @@ public function process(ContainerBuilder $container)
...
-        $classReflection = $subjectReflection = null;
 
         $excludes = $extensionMap['excludes'];
         unset($extensionMap['excludes']);
 
         foreach ($extensionMap as $type => $subjects) {
             foreach ($subjects as $subject => $extensionList) {
+                $class = null;
+
                 if ('admins' === $type) {
```
Comment (VincentLanglet) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:108]:
```suggestion
                if ('admins' === $type && $id === $subject) {
                    $extensions = array_merge($extensions, $extensionList);
                    
                    continue;
                }
                
                $class = $this->getManagedClass($admin, $container);
                if (!class_exists($class)) {
                    continue;
                }
```

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:108]: But it could be even better if `isExtensionClass` call `getManagedClass` to get the class.
This way you could write 
```
                if ('admins' === $type && $id === $subject || $this->isExtensionClass($type, $subject)) {
                    $extensions = array_merge($extensions, $extensionList);
                }
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 2 above)
Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: The original solution will be executed much faster if you use `instanceof` for an invalid class, because it will not check every other case.

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: What about a switch on `$type` ?

Comment (phansys) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: >The original solution will be executed much faster if you use instanceof for an invalid class, because it will not check every other case.

I don't know why, since in the proposed solution each `if` is checking first the type. If the type does not match, the condition is early evaluated as false (the `&&` is omitted) and the next `if` is executed.

Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: The original solution will fail fast, e.g. if the type will match. The other `elseif` cases would not be evaluated
```
if ('instanceof' === $type) {
```

While this solution will evaluate every other `elseif` clause:
```
if ('instanceof' === $type && ($classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName())) {
```

Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: Same for this place: https://github.com/sonata-project/SonataAdminBundle/pull/6028/files#r403769115

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: Still, 
```
if (something) {
  return true;
}

return false;
```
Is not great

```
return something;
```
Could be better.

WDYT about 
```
switch ($type) {
    case 'instanceof':
        return $classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName();
    case 'implements':
        return $classReflection->implementsInterface($subject);
    case 'extends':
        return $classReflection->isSubclassOf($subject);
    case 'uses':
        return $this->hasTrait($classReflection, $subject);
}
```

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:233]: And for https://github.com/sonata-project/SonataAdminBundle/pull/6028/files#r403769115

You can do 
```
if ('admins' === $type) {
    if ($id === $subject) {
        $extensions = array_merge($extensions, $extensionList);
    }
                    
    continue;
}
                
$class = $this->getManagedClass($admin, $container);
if (class_exists($class) && $this->isExtensionClass($type, $class, $subject)) {
    $extensions = array_merge($extensions, $extensionList);
}
```

It fails as fast as the original version

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 3 above)
Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:107]: > But it could be even better if `isExtensionClass` call `getManagedClass` to get the class.

I plan to reuse the `isExtensionClass` with the next PR.

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 4]
```diff
@@ -156,7 +136,23 @@ protected function getExtensionsForAdmin($id, Definition $admin, ContainerBuilde
      */
     protected function getManagedClass(Definition $admin, ContainerBuilder $container)
     {
-        return $container->getParameterBag()->resolveValue($admin->getArgument(1));
+        $argument = $admin->getArgument(1);
+        $class = $container->getParameterBag()->resolveValue($argument);
+
+        if (null === $class) {
+            throw new \DomainException(sprintf('The admin "%s" does not have a valid manager', $admin->getClass()));
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:143]:
```suggestion
            throw new \DomainException(sprintf('The admin "%s" does not have a valid manager.', $admin->getClass()));
```

**Comment by phansys** [comment]
Please, check [flintci/pr](https://flintci.io/repositories/61/analyses/37336).

File: tests/DependencyInjection/Compiler/ExtensionCompilerPassTest.php [hunk 5]
```diff
@@ -42,13 +42,11 @@
 class ExtensionCompilerPassTest extends TestCase
 {
     /**
-     * @var SonataAdminExtension
-     */
+     * @var SonataAdminExtension */
```
Comment (VincentLanglet) [inline tests/DependencyInjection/Compiler/ExtensionCompilerPassTest.php:45]: Why these changes.

This seems weird. half mutiline, half singleline.

File: tests/DependencyInjection/Compiler/ExtensionCompilerPassTest.php [hunk 6]
```diff
@@ -42,13 +42,11 @@
...
     private $extension;
 
     /**
-     * @var array
-     */
+     * @var array */
```
Comment (VincentLanglet) [inline tests/DependencyInjection/Compiler/ExtensionCompilerPassTest.php:49]: Same

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 7]
```diff
@@ -157,7 +137,23 @@ protected function getExtensionsForAdmin($id, Definition $admin, ContainerBuilde
...
+        }
+
+        if (!\is_string($class)) {
+            throw new \TypeError(sprintf(
+                'Argument "%s" for admin class "%s" must be of type string, %s given.',
+                $argument,
+                $admin->getClass(),
+                \is_object($class) ? \get_class($class) : \gettype($class)
+            ));
+        }
```
Comment (VincentLanglet) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:154]: Adding these Exception/Error aren't BC-break ?

Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:154]: This should not be a BC break, because the old code should have failed here if you did not provide a valid string:

https://github.com/sonata-project/SonataAdminBundle/pull/6028/files#diff-9dcbafe8f12838a40a139ccaa444c466L113

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 8]
```diff
@@ -97,48 +97,28 @@ public function process(ContainerBuilder $container)
...
         foreach ($extensionMap as $type => $subjects) {
             foreach ($subjects as $subject => $extensionList) {
+                $class = null;
+
                 if ('admins' === $type) {
                     if ($id === $subject) {
                         $extensions = array_merge($extensions, $extensionList);
                     }
                 } else {
                     $class = $this->getManagedClass($admin, $container);
```
Comment (VincentLanglet) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:113]:
```suggestion
                if ('admins' === $type) {
                    if ($id === $subject) {
                        $extensions = array_merge($extensions, $extensionList);
                    }
                    
                    continue;
                }
                
                $class = $this->getManagedClass($admin, $container);
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 7 above)
Comment (jordisala1991) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:154]: I dont see consistency here. So this is not BC break, but this other is? 

https://github.com/sonata-project/sonata-doctrine-extensions/pull/184#discussion_r399815034

**Review by VincentLanglet — DISMISSED** [review]

**Comment by core23** [comment]
Please review again @jordisala1991

**Review by jordisala1991 — DISMISSED** [review]

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 9]
```diff
@@ -215,6 +228,25 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
         return $this->hasTrait($parentClass, $traitName);
     }
 
+    private function isExtensionClass(string $type, string $class, string $subject): bool
+    {
+        $classReflection = new \ReflectionClass($class);
+        $subjectReflection = new \ReflectionClass($subject);
+
+        switch ($type) {
+            case 'instanceof':
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:237]:
```suggestion

        switch ($type) {
            case 'instanceof':
                $subjectReflection = new \ReflectionClass($subject);
```

**Review by VincentLanglet — APPROVED** [review]

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 7 above)
Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php]: Changed this, please review again @jordisala1991

**Review by VincentLanglet — APPROVED** [review]

**Review by franmomu — APPROVED** [review]

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 10]
```diff
@@ -215,6 +235,30 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
         return $this->hasTrait($parentClass, $traitName);
     }
 
+    /**
+     * @phpstan-param class-string $class
+     * @phpstan-param class-string $subject
+     */
+    private function isExtensionClass(string $type, string $class, string $subject): bool
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:242]: Taking a look about the responsibility of this method, maybe we could find a better name. It is not only checking if a class is of a given type (which could be simplified with `is_a($class, $subject, true)`), but this is checking also the inheritance type.
Additionally, I think we should rethink the argument names (and their order), since we are checking if a `$class` is an instance of `$type` with the expected `$inheritance`.
I'm proposing something like this:
```
isInstanceOf(string $class, string $type, string $inheritance)
```

Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php]: Change it to `isSubtypeOf`, because `instanceof` is a reserved keyword and could be confusing.

**Review by VincentLanglet — APPROVED** [review]

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 11]
```diff
@@ -153,15 +134,52 @@ protected function getExtensionsForAdmin($id, Definition $admin, ContainerBuilde
...
+        }
+
+        return $class;
     }
 
     /**
-     * @return array an array with the following structure.
+     * @param array<string, array<string, array<string, string>>> $config
+     *
+     * @return array<string, array<string, array<string, array<string, array<string, string>>>>> an array with the following structure.
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:182]:
```suggestion
     * @phpstan-return array<'excludes'|'admins'|'implements'|'extends'|'instanceof'|'uses', array<
     *                   string, array<string, array<string, array{priority, int}>>
     *               >>
     * @return array<string, array<string, array<string, array<string, array<string, int>>>>> an array with the 
```

or maybe both modulo some prefix?

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 12]
```diff
@@ -215,6 +235,30 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
         return $this->hasTrait($parentClass, $traitName);
     }
 
+    /**
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:238]:
```suggestion
    /**
     * @phpstan-param 'excludes'|'admins'|'implements'|'extends'|'instanceof'|'uses' $type
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 11 above)
Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:182]: I'm not sure phpstan supports this syntax.

Comment (greg0ire) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:182]: Which part?

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:182]: `'excludes'|'admins'|'implements'|'extends'|'instanceof'|'uses'`

I was wrong https://phpstan.org/writing-php-code/phpdoc-types#literals-and-constants

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 10 above)
Comment (phansys) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php]: >Change it to isSubtypeOf, because instanceof is a reserved keyword and could be confusing.

Good, but please, note that I proposed `isInstanceOf()`, not `instanceof()`.
By the other hand, I still think the main concept of this method hasn't the right approach, conceptually speaking.
I think we are checking if a **subject** is subtype of a given **class**, but it currently is expressed as if we are checking if a **class** is a subtype of a given **subject**.

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 12 above)
Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:238]: phpstan can't understand this:

```
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   DependencyInjection/Compiler/ExtensionCompilerPass.php
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  121    Parameter #1 $type of method Sonata\AdminBundle\DependencyInjection\Compiler\ExtensionCompilerPass::isSubtypeOf() expects 'admins'|'excludes'|'extends'|'implements'|'instanceof'|'uses', string given.
  219    Method Sonata\AdminBundle\DependencyInjection\Compiler\ExtensionCompilerPass::flattenExtensionConfiguration() should return array<'admins'|'excludes'|'extends'|'implements'|'instanceof'|'uses',
         array<string, array<string, array<string, array('priority' => string)>>>> but returns array<string, array<string, array<string, array('priority' => array<string, string>)>>>.
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:238]: It's because `type` is considered as string, because of 
```
@param array<string, array<string, array<string, array<string, array<string, mixed>>>>> $extensionMap
```

Maybe it's better to keep string, since `isSubtypeOf` does support every possible string ; there is a default `return false`.

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 13]
```diff
@@ -215,6 +235,30 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
...
+                return $classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName();
+            case 'implements':
+                return $classReflection->implementsInterface($subject);
+            case 'extends':
+                return $classReflection->isSubclassOf($subject);
+            case 'uses':
+                return $this->hasTrait($classReflection, $subject);
+        }
+
+        return false;
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:259]: In case `$type` could be optional:
```suggestion
        return is_a($class, $subject, true);
```

Comment (core23) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:259]: This is not the purpose of the method, also type is a required field

Comment (phansys) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php:259]: I'm aware that `$type` argument is currently required, that's why I left this suggestion indicating this change could be possible if we make it optional.

**Comment by VincentLanglet** [comment]
@phansys Is this PR ok for you ? I'm missing the blocking point.

**Comment by phansys** [comment]
> @phansys Is this PR ok for you ? I'm missing the blocking point.

I'd not say this is a blocker, but a personal concern about the inheritance concept and its operands.
IMO, the argument 1 we pass to [`ReflectionClass::isSubclassOf()`](https://www.php.net/manual/en/reflectionclass.issubclassof.php) and other reflection methods should not be named `$subject`. My proposal is to rename this argument to `$class` or something else, since from my POV, the subject is actually the argument 2.
**CURRENT SIGNATURE**:
```php
private function isSubtypeOf(string $type, string $class, string $subject): bool
// $class->isSubclassOf($subject)
```
**PROPOSED SIGNATURE**:
```php
private function isSubtypeOf(string $type, string $subject, string $class): bool
// $subject->isSubclassOf($class)
```

**Comment by VincentLanglet** [comment]
@core23 Is it ok for you to change the signature ?

**Comment by core23** [comment]
Sorry for the delay. Please review again

**Review by phansys — APPROVED** [review]
I still think that https://github.com/sonata-project/SonataAdminBundle/pull/6028#issuecomment-695843706  fits better for the responsibilities of these arguments, but I don't want to block this PR for more time.

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @core23
